lib/mllp/client.ex

```diff
@@ -356,282 +377,363 @@ defmodule MLLP.Client do
+  end
+
+  def receiving(:info, {transport, socket, incoming}, %{socket: socket} = data)
+      when transport in [:tcp, :ssl] do
+    new_data = handle_received(incoming, data)
+    next_state = (new_data.caller && :receiving) || :connected
+    {:next_state, next_state, new_data}
+  end
+
+  def receiving(:info, {transport_closed, socket}, %{socket: socket} = data)
+      when transport_closed in [:tcp_closed, :ssl_closed] do
+    {:next_state, :disconnected, handle_closed(data)}
+  end
 
-        case new_buf do
-          <<@header, _ack::binary-size(check), @trailer>> ->
-            {:ok, new_buf}
+  def receiving(:info, {transport_error, socket, reason}, %{socket: socket} = data)
+      when transport_error in [:tcp_error, :ssl_error] do
+    {:next_state, :disconnected, handle_error(reason, maybe_close(reason, data))}
+  end
 
-          <<@header, _rest::binary>> ->
-            time_credit = update_recv_time_credit(time_left, time_owed + elapsed)
-            recv_ack(state, time_credit, new_buf)
+  def receiving(event, unknown, _data) do
+    unexpected_message(:receiving, event, unknown)
+  end
 
-          _ ->
-            {:error, :invalid_reply}
-        end
+  ########################################
+  ### End of :gen_statem callbacks ###
+  ########################################
 
-      {:error, _} = err ->
-        err
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
+
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = data) do
+    ## No caller, ignore
+    data
+  end
+
+  defp handle_received(<<@header, _rest::binary>> = reply, data) do
+    receive_impl(reply, data)
+  end
+
+  ## No header in the first packet
+  defp handle_received(_reply, %{receive_buffer: []} = data) do
+    reply_to_caller({:error, :invalid_reply}, data)
+  end
+
+  ## The rest of MLLP (after the header was received)
+  defp handle_received(reply, data) do
+    receive_impl(reply, data)
+  end
+
+  defp receive_impl(reply, %{receive_buffer: buffer} = data) do
+    new_buf = update_receive_buffer(buffer, reply)
+
+    case trailer_check(reply) do
+      :data_after_trailer ->
+        Logger.error("Client #{inspect(self())} received data following the trailer")
+        reply_to_caller({:error, :data_after_trailer}, data)
+
+      true ->
+        Logger.debug("Client #{inspect(self())} received a full MLLP!")
+        reply_to_caller({:ok, buffer_to_binary(new_buf)}, data)
+
+      false ->
+        Logger.debug("Client #{inspect(self())} received a MLLP fragment: #{reply}")
+        Map.put(data, :receive_buffer, new_buf)
     end
   end
 
-  defp do_recv(state, length, :infinity) do
-    res = state.tcp.recv(state.socket, length, :infinity)
-    {res, 0}
+  defp trailer_check(packet) do
```

Comment:
Interestingly, this handles an edge case, but not the one I was referring to :)  I haven't seen this one in production, but it could happen. 

Back to the original edge case first that I believe I wasn't clear about : 

The trailer consists of two bytes, the file separator byte and a byte for a carriage return. What can and does happen is you get everything in chunks (in various ways) all the way to the file separator, then the carriage return arrives on its own. 

As an example, you get the first chunk it contains the header, followed by some hl7 data, then maybe you get a second chunk it contains more hl7 data and the first byte of the trailer, and finally you get carriage return in its own "packet". 

As for the binary match approach, I'm not sure on that vs a straight pattern match. I would ask @josevalim for his opinion as well.

Reply:
> Interestingly, this handles an edge case, but not the one I was referring to :) 

I think it does handle that case, too :-)
The code waits for a full trailer, and it won't matter if the carriage return comes separately.
The test case for this "last byte" is a bit trickier to make, but I'll spend some time putting it together.

Reply:
I meant creating a unit test for that case is tricky, we'd need to do a mock that mimics that exact "last byte" case, or alternatively, expose the "parsing" logic as a public fun and write the test for it.

Reply:
A mock would be fine, since that's already in use. I'd also be quite ok with exposing the parsing logic (I've done precisely this in a previous mllp related project).